libnm-core/nm-setting-match.c

```diff
@@ -0,0 +1,306 @@
+{
+	g_return_if_fail (NM_IS_SETTING_MATCH (setting));
+
+	if (setting->interface_name->len != 0) {
+		g_ptr_array_set_size (setting->interface_name, 0);
+		g_object_notify (G_OBJECT (setting), NM_SETTING_MATCH_INTERFACE_NAME);
+	}
+}
+
+/**
+ * nm_setting_match_get_interface_names:
+ * @setting: the #NMSettingMatch
+ *
+ * Returns all the interface names.
+ *
+ * Returns: (transfer none): the configured interface names.
+ *
+ * Since: 1.14
+ **/
+const char *const *
+nm_setting_match_get_interface_names (NMSettingMatch *setting, guint *length)
+{
+	g_return_val_if_fail (NM_IS_SETTING_MATCH (setting), NULL);
+	g_return_val_if_fail (length, NULL);
+
+	NM_SET_OUT (length, setting->interface_name->len);
+	return (const char *const *) setting->interface_name->pdata;
+}
+
+static gboolean
+verify (NMSetting *setting, NMConnection *connection, GError **error)
+{
+	/* TODO */
+
+	return TRUE;
+}
+
+static void
+set_property (GObject *object, guint prop_id,
+              const GValue *value, GParamSpec *pspec)
+{
+	NMSettingMatch *self = NM_SETTING_MATCH (object);
+
+	switch (prop_id) {
+	case PROP_INTERFACE_NAME:
+		g_ptr_array_unref (self->interface_name);
+		self->interface_name = _nm_utils_strv_to_ptrarray (g_value_get_boxed (value));
+		break;
+	default:
+		G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+		break;
+	}
+}
+
+static void
+get_property (GObject *object, guint prop_id,
+              GValue *value, GParamSpec *pspec)
+{
+	NMSettingMatch *self = NM_SETTING_MATCH (object);
+
+	switch (prop_id) {
+	case PROP_INTERFACE_NAME:
+		g_value_take_boxed (value, _nm_utils_ptrarray_to_strv (self->interface_name));
+		break;
+	default:
+		G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+		break;
+	}
+}
+
+static void
+nm_setting_match_init (NMSettingMatch *setting)
+{
+	setting->interface_name = g_ptr_array_new_with_free_func (g_free);
+}
+
+static void
+finalize (GObject *object)
+{
+	NMSettingMatch *self = NM_SETTING_MATCH (object);
+
+	g_ptr_array_unref (self->interface_name);
+
+	G_OBJECT_CLASS (nm_setting_match_parent_class)->finalize (object);
+}
+
+static void
+nm_setting_match_class_init (NMSettingMatchClass *setting_class)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS (setting_class);
+	NMSettingClass *parent_class = NM_SETTING_CLASS (setting_class);
+
+	object_class->finalize         = finalize;
+	object_class->get_property     = get_property;
+	object_class->set_property     = set_property;
+	parent_class->verify           = verify;
+
+
+	/**
+	 * NMSettingMatch:interface-name
+	 *
+	 * A list of interface names to match. Each element is a shell wildcard
+	 * pattern.  When an element is prefixed with exclamation mark (!) the
+	 * condition is inverted.
+	 *
+	 * A candidate interface name is considered matching when both these
+	 * conditions are satisfied: (a) any of the elements not prefixed with '!'
+	 * matches or there aren't such elements; (b) none of the elements
+	 * prefixed with '!' match.
```

Comment:
The idea is that, like it happens now for con.interface-name and
wired.mac, all properties must be matching for the connection to
match. That is, the result is the logical AND of all conditions, where
empty properties are considered as a match.

Since there aren't any other match properties at the moment, I don't
know if it's a good idea to explain it in the documentation.

I think this is the most useful behavior; systemd-networkd uses the
same approach, and we already do it for the name and mac non-wildcard
properties. And it is also easier to implement because we must only
check each property separately and bail out when one does not match.